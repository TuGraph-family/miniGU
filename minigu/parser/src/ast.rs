pub mod catalog;
pub mod common;
pub mod data;
pub mod element;
pub mod expression;
pub mod lexical;
pub mod object;
pub mod predicate;
pub mod procedure;
pub mod program;
pub mod query;
pub mod reference;
pub mod session;
pub mod transaction;

pub use catalog::*;
pub use common::*;
pub use data::*;
pub use element::*;
pub use expression::*;
pub use lexical::*;
pub use object::*;
pub use predicate::*;
pub use procedure::*;
pub use program::*;
pub use query::*;
pub use reference::*;
pub use session::*;
pub use transaction::*;
