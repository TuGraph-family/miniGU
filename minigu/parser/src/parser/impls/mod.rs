mod catalog;
mod common;
mod data;
mod lexical;
mod object_expr;
mod object_ref;
mod predicate;
mod procedure_call;
mod procedure_spec;
mod program;
mod query;
mod session;
mod transaction;
mod type_element;
mod value_expr;
mod variable;

pub use program::gql_program;
