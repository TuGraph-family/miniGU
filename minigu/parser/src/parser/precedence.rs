use winnow::combinator::opt;
use winnow::error::{ErrMode, ParserError};
use winnow::stream::{Stream, StreamIsPartial};
use winnow::{ModalParser, ModalResult, Parser};

/// Specifies the associativity of an (infix) operator.
#[derive(Debug, Clone, Copy)]
pub(super) enum Assoc {
    Left,
    Right,
    Neither,
}

type Precedence = i64;
type UnaryCallback<I, Operand, E> = fn(&mut I, Operand) -> ModalResult<Operand, E>;
type BinaryCallback<I, Operand, E> = fn(&mut I, Operand, Operand) -> ModalResult<Operand, E>;

/// An operator precedence parser based on the [Pratt algorithm](https://en.wikipedia.org/wiki/Operator-precedence_parser).
/// The implementation is adapted from https://github.com/winnow-rs/winnow/pull/614.
pub(super) fn precedence<I, ParseOperand, ParseInfix, ParsePrefix, ParseSuffix, Operand, E>(
    init: Precedence,
    mut operand: ParseOperand,
    mut prefix: ParsePrefix,
    mut suffix: ParseSuffix,
    mut infix: ParseInfix,
) -> impl ModalParser<I, Operand, E>
where
    I: Stream + StreamIsPartial,
    ParseOperand: ModalParser<I, Operand, E>,
    ParseInfix: ModalParser<I, (Assoc, Precedence, BinaryCallback<I, Operand, E>), E>,
    ParsePrefix: ModalParser<I, (Precedence, UnaryCallback<I, Operand, E>), E>,
    ParseSuffix: ModalParser<I, (Precedence, UnaryCallback<I, Operand, E>), E>,
    E: ParserError<I>,
{
    move |i: &mut I| precedence_impl(i, &mut operand, &mut prefix, &mut suffix, &mut infix, init)
}

fn precedence_impl<I, ParseOperand, ParseInfix, ParsePrefix, ParseSuffix, Operand, E>(
    i: &mut I,
    parse_operand: &mut ParseOperand,
    prefix: &mut ParsePrefix,
    suffix: &mut ParseSuffix,
    infix: &mut ParseInfix,
    min_power: Precedence,
) -> ModalResult<Operand, E>
where
    I: Stream + StreamIsPartial,
    ParseOperand: ModalParser<I, Operand, E>,
    ParseInfix: ModalParser<I, (Assoc, Precedence, BinaryCallback<I, Operand, E>), E>,
    ParsePrefix: ModalParser<I, (Precedence, UnaryCallback<I, Operand, E>), E>,
    ParseSuffix: ModalParser<I, (Precedence, UnaryCallback<I, Operand, E>), E>,
    E: ParserError<I>,
{
    let operand = opt(parse_operand.by_ref()).parse_next(i)?;
    let mut operand = if let Some(operand) = operand {
        operand
    } else {
        let len = i.eof_offset();
        let (power, fold_prefix) = prefix.parse_next(i)?;
        if i.eof_offset() == len {
            return Err(ErrMode::assert(i, "`prefix` parsers must always consume"));
        }
        let operand = precedence_impl(i, parse_operand, prefix, suffix, infix, power)?;
        fold_prefix(i, operand)?
    };

    let mut prev_op_is_neither = None;
    while i.eof_offset() > 0 {
        let start = i.checkpoint();
        if let Some((power, fold_suffix)) = opt(suffix.by_ref()).parse_next(i)? {
            if power < min_power {
                i.reset(&start);
                break;
            }
            operand = fold_suffix(i, operand)?;
            continue;
        }

        let start = i.checkpoint();
        let parse_result = opt(infix.by_ref()).parse_next(i)?;
        if let Some((assoc, power, fold_infix)) = parse_result {
            let mut is_neither = None;
            let (lpower, rpower) = match assoc {
                Assoc::Left => (power, power + 1),
                Assoc::Right => (power, power - 1),
                Assoc::Neither => {
                    is_neither = Some(power);
                    (power, power + 1)
                }
            };
            if lpower < min_power || prev_op_is_neither.is_some_and(|p| lpower == p) {
                i.reset(&start);
                break;
            }
            prev_op_is_neither = is_neither;
            let rhs = precedence_impl(i, parse_operand, prefix, suffix, infix, rpower)?;
            operand = fold_infix(i, operand, rhs)?;
            continue;
        }
        break;
    }
    Ok(operand)
}

#[cfg(test)]
mod tests {
    use core::str::FromStr;

    use winnow::ascii::{digit1, multispace0};
    use winnow::combinator::{delimited, dispatch, empty, fail, peek};
    use winnow::token::any;

    use super::*;

    fn operand(i: &mut &str) -> ModalResult<i64> {
        digit1.try_map(i64::from_str).parse_next(i)
    }

    fn parenthesized(i: &mut &str) -> ModalResult<i64> {
        delimited('(', expr, ')').parse_next(i)
    }

    fn expr(i: &mut &str) -> ModalResult<i64> {
        precedence(
            0,
            delimited(multispace0, dispatch!{
                peek(any);
                '(' => parenthesized,
                '0'..='9' => operand,
                _ => fail
            }, multispace0),
            delimited(
                multispace0,
                dispatch! {
                    any;
                    '+' => empty.value((2, (|_: &mut _, a: i64| Ok(a)) as _)),
                    '-' => empty.value((2, (|_: &mut _, a: i64| Ok(-a)) as _)),
                    _ => fail
                },
                multispace0,
            ),
            delimited(
                multispace0,
                dispatch! {
                    any;
                    '!' => empty.value((3, (|_: &mut _, a: i64| Ok((1..=a).product())) as _)),
                    _ => fail
                },
                multispace0,
            ),
            delimited(multispace0, dispatch! {
                any;
                '+' => empty.value((Assoc::Left, 0, (|_: &mut _, a: i64, b: i64| Ok(a + b)) as _)),
                '-' => empty.value((Assoc::Left, 0, (|_: &mut _, a: i64, b: i64| Ok(a - b)) as _)),
                '*' => empty.value((Assoc::Left, 1, (|_: &mut _, a: i64, b: i64| Ok(a * b)) as _)),
                '/' => empty.value((Assoc::Left, 1, (|_: &mut _, a: i64, b: i64| Ok(a / b)) as _)),
                _ => fail
            }, multispace0),
        )
        .parse_next(i)
    }

    #[test]
    fn test_parse_operand() {
        assert_eq!(operand.parse("123").unwrap(), 123);
    }

    #[test]
    fn test_simple_expr() {
        assert_eq!(expr.parse("1 +  3* 4 /2 ").unwrap(), 7);
    }

    #[test]
    fn test_complex_expr() {
        assert_eq!(expr.parse("3- (+ 1 * - (3 +4) + 10)!").unwrap(), -3);
    }
}
