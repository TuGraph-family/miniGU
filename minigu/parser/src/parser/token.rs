use logos::Logos;
use winnow::error::ParserError;
use winnow::stream::{ContainsToken, Stream, StreamIsPartial, TokenSlice};
use winnow::token::one_of;
use winnow::{Parser, Stateful};

use crate::lexer::{LexerError, TokenKind};
use crate::span::Span;

#[derive(Debug, Clone, PartialEq, Eq)]
pub(super) struct Token<'a> {
    pub(super) kind: TokenKind<'a>,
    pub(super) slice: &'a str,
    pub(super) span: Span,
}

impl<'a> Token<'a> {
    #[inline(always)]
    pub(super) fn kind(&self) -> &TokenKind<'a> {
        &self.kind
    }

    #[inline(always)]
    pub(super) fn span(&self) -> Span {
        self.span
    }
}

impl<'a> ContainsToken<&Token<'a>> for &TokenKind<'a> {
    #[inline(always)]
    fn contains_token(&self, token: &Token<'a>) -> bool {
        token.kind.eq(self)
    }
}

impl<I, E> Parser<I, I::Token, E> for TokenKind<'_>
where
    I: Stream + StreamIsPartial,
    I::Token: Clone,
    E: ParserError<I>,
    for<'b> &'b Self: ContainsToken<I::Token>,
{
    fn parse_next(&mut self, i: &mut I) -> Result<I::Token, E> {
        one_of(&*self).parse_next(i)
    }
}

#[derive(Debug, Default)]
pub(super) struct State {
    recursion: usize,
}

pub(super) type TokenStream<'a, 'b> = Stateful<TokenSlice<'b, Token<'a>>, State>;

pub(super) fn tokenize(input: &str) -> Result<Vec<Token<'_>>, LexerError> {
    let mut lexer = TokenKind::lexer(input).spanned();
    let mut tokens = Vec::new();
    while let Some((kind, span)) = lexer.next() {
        let kind = kind?;
        let slice = lexer.slice();
        let span = span.try_into().map_err(|_| LexerError::InvalidToken)?;
        tokens.push(Token { kind, slice, span });
    }
    Ok(tokens)
}

pub(super) fn build_token_stream<'a, 'b>(input: &'b [Token<'a>]) -> TokenStream<'a, 'b> {
    let input = TokenSlice::new(input);
    Stateful {
        input,
        state: State::default(),
    }
}
