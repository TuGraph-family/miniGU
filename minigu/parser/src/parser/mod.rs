pub use options::ParseOptions;
pub use token::{Token, tokenize, tokenize_full};

use crate::ast::Program;
use crate::error::Error;
use crate::span::Spanned;

mod impls;
mod options;
mod precedence;
mod token;
mod utils;

/// Parses a GQL query into a spanned abstract syntax tree with default options.
///
/// See [`ParseOptions`] for more information on how to configure the parser.
///
/// # Errors
///
/// This function will return an error if `gql` is not a valid GQL query. The error will carry
/// fancy diagnostics if feature `miette` is enabled.
///
/// # Examples
///
/// ```
/// # use gql_parser::parse_gql;
/// let program = parse_gql("SESSION CLOSE");
/// assert!(program.is_ok());
/// assert_eq!(program.unwrap().span(), 0..13);
/// ```
pub fn parse_gql(gql: &str) -> Result<Spanned<Program>, Error> {
    ParseOptions::new().parse(gql)
}
