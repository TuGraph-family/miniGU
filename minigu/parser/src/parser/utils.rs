use winnow::Parser;
use winnow::combinator::{Repeat, peek};
use winnow::error::ParserError;
use winnow::stream::{Location, Stream, StreamIsPartial};
use winnow::token::any;

use super::token::Token;
use crate::lexer::TokenKind;
use crate::span::Spanned;

/// A helper macro for executing a parser on an input string. **This should only be used in unit
/// tests!**
#[cfg(all(test, feature = "serde"))]
macro_rules! parse {
    ($parser:expr, $input:literal) => {{
        let input = $input;
        $crate::parser::token::tokenize(input)
            .ok()
            .and_then(|tokens| {
                let stream = crate::parser::token::build_token_stream(&tokens, Default::default());
                $parser.parse(stream).ok()
            })
    }};
}

#[cfg(all(test, feature = "serde"))]
pub(super) use parse;

/// A helper trait for wrapping the output in [`Spanned`].
pub(super) trait ToSpanned<I, O, E>: Parser<I, O, E> {
    /// Wraps the output in [`Spanned`].
    #[inline(always)]
    fn spanned(self) -> impl Parser<I, Spanned<O>, E>
    where
        Self: Sized,
        I: Stream + Location,
    {
        self.with_span().map(|(inner, span)| Spanned(inner, span))
    }
}

/// A helper trait for operating on the value wrapped in [`Spanned`].
pub(super) trait SpannedParserExt<I, O, E>: Parser<I, Spanned<O>, E> {
    /// Unwraps the output from [`Spanned`].
    #[inline(always)]
    fn unspanned(self) -> impl Parser<I, O, E>
    where
        Self: Sized,
    {
        self.map(|spanned| spanned.0)
    }

    /// Maps the inner value of the output while keeping the span unchanged.
    #[inline(always)]
    fn map_inner<F, O2>(self, mut f: F) -> impl Parser<I, Spanned<O2>, E>
    where
        Self: Sized,
        F: FnMut(O) -> O2,
    {
        self.map(move |Spanned(inner, span)| Spanned(f(inner), span))
    }

    /// Updates the span of the output while keeping the inner value unchanged.
    #[inline(always)]
    fn update_span(self) -> impl Parser<I, Spanned<O>, E>
    where
        Self: Sized,
        I: Stream + Location,
    {
        self.with_span().map(|(mut output, new_span)| {
            output.1 = new_span;
            output
        })
    }
}

impl<I, O, E, P> ToSpanned<I, O, E> for P where P: Parser<I, O, E> {}

impl<I, O, E, P> SpannedParserExt<I, O, E> for P where P: Parser<I, Spanned<O>, E> {}

/// A helper parser for taking one token ahead.
#[inline(always)]
pub(super) fn take1<'a: 'b, 'b, I, E>(input: &mut I) -> Result<&'b TokenKind<'a>, E>
where
    I: Stream<Token = &'b Token<'a>> + StreamIsPartial,
    E: ParserError<I>,
{
    any.map(|t: &Token| &t.kind).parse_next(input)
}

/// A helper parser for taking two tokens ahead.
#[inline(always)]
pub(super) fn take2<'a: 'b, 'b, I, E>(
    input: &mut I,
) -> Result<(&'b TokenKind<'a>, &'b TokenKind<'a>), E>
where
    I: Stream<Token = &'b Token<'a>> + StreamIsPartial,
    E: ParserError<I>,
{
    (any, any)
        .map(|(t1, t2): (&Token, &Token)| (&t1.kind, &t2.kind))
        .parse_next(input)
}

/// A helper parser for peeking one lookahead token.
#[inline(always)]
pub(super) fn peek1<'a: 'b, 'b, I, E>(input: &mut I) -> Result<&'b TokenKind<'a>, E>
where
    I: Stream<Token = &'b Token<'a>> + StreamIsPartial,
    E: ParserError<I>,
{
    peek(any).map(|t: &Token| &t.kind).parse_next(input)
}

/// A helper parser for peeking two lookahead tokens.
#[inline(always)]
pub(super) fn peek2<'a: 'b, 'b, I, E>(
    input: &mut I,
) -> Result<(&'b TokenKind<'a>, &'b TokenKind<'a>), E>
where
    I: Stream<Token = &'b Token<'a>> + StreamIsPartial,
    E: ParserError<I>,
{
    peek((any, any))
        .map(|(t1, t2): (&Token, &Token)| (&t1.kind, &t2.kind))
        .parse_next(input)
}

/// A helper parser for peeking three lookahead tokens.
#[inline(always)]
pub(super) fn peek3<'a: 'b, 'b, I, E>(
    input: &mut I,
) -> Result<(&'b TokenKind<'a>, &'b TokenKind<'a>, &'b TokenKind<'a>), E>
where
    I: Stream<Token = &'b Token<'a>> + StreamIsPartial,
    E: ParserError<I>,
{
    peek((any, any, any))
        .map(|(t1, t2, t3): (&Token, &Token, &Token)| (&t1.kind, &t2.kind, &t3.kind))
        .parse_next(input)
}

/// A helper macro for defining parser aliases.
macro_rules! def_parser_alias {
    ($name:ident, $parser:ident, $output:ty) => {
        #[inline(always)]
        pub fn $name(
            input: &mut $crate::parser::token::TokenStream,
        ) -> winnow::ModalResult<$output> {
            $parser(input)
        }
    };
}

pub(super) use def_parser_alias;
