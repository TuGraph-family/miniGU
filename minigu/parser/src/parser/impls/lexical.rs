use winnow::combinator::{cut_err, dispatch, empty, fail, peek};
use winnow::token::one_of;
use winnow::{ModalResult, Parser};

use crate::ast::{
    BooleanLiteral, Ident, Literal, StringLiteral, StringLiteralKind, UnsignedInteger,
    UnsignedIntegerKind, UnsignedNumericLiteral,
};
use crate::lexer::TokenKind;
use crate::parser::token::{Token, TokenStream, any};
use crate::parser::utils::{SpannedParserExt, ToSpanned, def_parser_alias};
use crate::span::Spanned;

pub fn regular_identifier(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    winnow::token::any
        .verify_map(|token: &Token| match &token.kind {
            &TokenKind::RegularIdentifier(name) => Some(name.into()),
            kind if kind.is_non_reserved_word() => Some(token.slice.into()),
            _ => None,
        })
        .spanned()
        .parse_next(input)
}

pub fn delimited_identifier(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    dispatch! {any;
        TokenKind::AccentQuoted(quoted) | TokenKind::DoubleQuoted(quoted) => {
            cut_err(empty.verify_map(|_| quoted.unescape()))
        },
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn identifier(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    dispatch! {peek(any);
        TokenKind::RegularIdentifier(_) => regular_identifier,
        kind if kind.is_non_reserved_word() => regular_identifier,
        TokenKind::AccentQuoted(_) | TokenKind::DoubleQuoted(_) => delimited_identifier,
        _ => fail
    }
    .parse_next(input)
}

def_parser_alias!(object_name, identifier, Spanned<Ident>);
def_parser_alias!(
    object_name_or_binding_variable,
    regular_identifier,
    Spanned<Ident>
);
def_parser_alias!(directory_name, identifier, Spanned<Ident>);
def_parser_alias!(schema_name, identifier, Spanned<Ident>);
def_parser_alias!(graph_name, identifier, Spanned<Ident>);
def_parser_alias!(delimited_graph_name, delimited_identifier, Spanned<Ident>);
def_parser_alias!(graph_type_name, identifier, Spanned<Ident>);
def_parser_alias!(node_type_name, identifier, Spanned<Ident>);
def_parser_alias!(edge_type_name, identifier, Spanned<Ident>);
def_parser_alias!(binding_table_name, identifier, Spanned<Ident>);
def_parser_alias!(
    delimited_binding_table_name,
    delimited_identifier,
    Spanned<Ident>
);
def_parser_alias!(procedure_name, identifier, Spanned<Ident>);
def_parser_alias!(label_name, identifier, Spanned<Ident>);
def_parser_alias!(property_name, identifier, Spanned<Ident>);
def_parser_alias!(field_name, identifier, Spanned<Ident>);
def_parser_alias!(element_variable, binding_variable, Spanned<Ident>);
def_parser_alias!(path_variable, binding_variable, Spanned<Ident>);
def_parser_alias!(subpath_variable, regular_identifier, Spanned<Ident>);
def_parser_alias!(binding_variable, regular_identifier, Spanned<Ident>);

pub fn unsigned_literal(input: &mut TokenStream) -> ModalResult<Spanned<Literal>> {
    dispatch! {peek(any);
        TokenKind::True | TokenKind::False | TokenKind::Unknown => {
            boolean_literal.map_inner(Literal::Boolean)
        },
        TokenKind::Null => TokenKind::Null.value(Literal::Null).spanned(),
        TokenKind::DoubleQuoted(_) | TokenKind::SingleQuoted(_) => {
            character_string_literal.map_inner(Literal::String)
        },
        kind if kind.is_prefix_of_numeric_literal() => {
            unsigned_numeric_literal.map_inner(Literal::Numeric)
        },
        _ => fail,
    }
    .parse_next(input)
}

pub fn unsigned_numeric_literal(
    input: &mut TokenStream,
) -> ModalResult<Spanned<UnsignedNumericLiteral>> {
    unsigned_integer
        .map(UnsignedNumericLiteral::Integer)
        .spanned()
        .parse_next(input)
}

pub fn unsigned_integer(input: &mut TokenStream) -> ModalResult<Spanned<UnsignedInteger>> {
    dispatch! {any;
        &TokenKind::UnsignedDecimalInteger(integer) => empty
            .value(UnsignedInteger {
                kind: UnsignedIntegerKind::Decimal,
                integer: integer.into(),
            }),
        &TokenKind::UnsignedOctalInteger(integer) => empty
            .value(UnsignedInteger {
                kind: UnsignedIntegerKind::Octal,
                integer: integer.into(),
            }),
        &TokenKind::UnsignedHexInteger(integer) => empty
            .value(UnsignedInteger {
                kind: UnsignedIntegerKind::Hex,
                integer: integer.into(),
            }),
        &TokenKind::UnsignedBinaryInteger(integer) => empty
            .value(UnsignedInteger {
                kind: UnsignedIntegerKind::Binary,
                integer: integer.into(),
            }),
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn unsigned_decimal_integer(input: &mut TokenStream) -> ModalResult<Spanned<UnsignedInteger>> {
    winnow::token::any
        .verify_map(|token: &Token| match token.kind {
            TokenKind::UnsignedDecimalInteger(integer) => Some(UnsignedInteger {
                kind: UnsignedIntegerKind::Decimal,
                integer: integer.into(),
            }),
            _ => None,
        })
        .spanned()
        .parse_next(input)
}

pub fn boolean_literal(input: &mut TokenStream) -> ModalResult<Spanned<BooleanLiteral>> {
    dispatch! {any;
        TokenKind::True => empty.value(BooleanLiteral::True),
        TokenKind::False => empty.value(BooleanLiteral::False),
        TokenKind::Unknown => empty.value(BooleanLiteral::Unknown),
        _ => fail
    }
    .spanned()
    .parse_next(input)
}

pub fn general_parameter_reference(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    dispatch! {any;
        TokenKind::GeneralParameterReference(name) => {
            cut_err(empty.verify_map(|_| name.unescape()))
        },
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn substituted_parameter_reference(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    dispatch! {any;
        TokenKind::SubstitutedParameterReference(name) => {
            cut_err(empty.verify_map(|_| name.unescape()))
        },
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn character_string_literal(input: &mut TokenStream) -> ModalResult<Spanned<StringLiteral>> {
    dispatch! {any;
        TokenKind::SingleQuoted(quoted) | TokenKind::DoubleQuoted(quoted) => {
            cut_err(empty.verify_map(|_| {
                Some(StringLiteral {
                    kind: StringLiteralKind::Char,
                    literal: quoted.unescape()?,
                })
            }))
        },
        _ => fail
    }
    .spanned()
    .parse_next(input)
}

pub fn node_synonym(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Node, TokenKind::Vertex))
        .void()
        .parse_next(input)
}

pub fn edge_synonym(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Edge, TokenKind::Relationship))
        .void()
        .parse_next(input)
}

pub fn edges_synonym(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Edges, TokenKind::Relationships))
        .void()
        .parse_next(input)
}

pub fn implies(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Implies, TokenKind::RightDoubleArrow))
        .void()
        .parse_next(input)
}

#[cfg(all(test, feature = "serde"))]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;
    use crate::parser::utils::parse;

    #[test]
    fn test_identifier_1() {
        let parsed = parse!(identifier, "abcd");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_identifier_2() {
        let parsed = parse!(identifier, "`abcd`");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_identifier_3() {
        let parsed = parse!(identifier, "\"abcd\"");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_boolean_literal_1() {
        let parsed = parse!(boolean_literal, "true");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_boolean_literal_2() {
        let parsed = parse!(boolean_literal, "false");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_boolean_literal_3() {
        let parsed = parse!(boolean_literal, "unknown");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_literal_1() {
        let parsed = parse!(unsigned_literal, "false");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_literal_2() {
        let parsed = parse!(unsigned_literal, "null");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_character_string_literal() {
        let parsed = parse!(character_string_literal, "\"ab\\ncd\"");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_integer_1() {
        let parsed = parse!(unsigned_integer, "123");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_integer_2() {
        let parsed = parse!(unsigned_integer, "0o123");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_integer_3() {
        let parsed = parse!(unsigned_integer, "0x123");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_integer_4() {
        let parsed = parse!(unsigned_integer, "0b101");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_general_parameter_reference_1() {
        let parsed = parse!(general_parameter_reference, "$abc");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_general_parameter_reference_2() {
        let parsed = parse!(general_parameter_reference, "$\"abc\"");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_substituted_parameter_reference_1() {
        let parsed = parse!(substituted_parameter_reference, "$$abc");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_substituted_parameter_reference_2() {
        let parsed = parse!(substituted_parameter_reference, "$$\"abc\"");
        assert_yaml_snapshot!(parsed);
    }
}
