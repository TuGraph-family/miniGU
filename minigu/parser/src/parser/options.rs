use winnow::Parser;

use super::impls::gql_program;
use super::token::{build_token_stream, tokenize};
use crate::ast::Program;
use crate::error::Error;
use crate::span::Spanned;

/// Options which can be used to configure the behavior of the parser.
///
/// # Examples
/// Parsing a GQL query with default options:
/// ```no_run
/// use gql_parser::ParseOptions;
///
/// let parsed = ParseOptions::new().parse("match (: Person) -> (b: Person) return b");
/// ```
#[derive(Debug, Clone, Default)]
pub struct ParseOptions(ParseOptionsInner);

impl ParseOptions {
    /// Create a default set of parse options for configuration.  
    ///
    /// # Examples
    /// ```no_run
    /// use gql_parser::ParseOptions;
    ///
    /// let mut options = ParseOptions::new();
    /// let parsed = options.unescape(true).parse("CREATE GRAPH mygraph ANY");
    /// ```
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Sets whether quoted character sequences should be unescaped by the parser.
    ///
    /// If set to `true` (default), the parser will unescape quoted sequences in the input query.
    /// For example, `""` in GQL is unescaped to `"` in a double quoted character sequence.
    ///
    /// Otherwise, the parser will leave the raw strings unchanged in the output, and the
    /// caller should handle them manually.
    ///
    /// # Examples
    /// // TODO: Fill this part.
    ///
    /// Parsing a GQL query with quoted character sequences unescaped:
    /// ```no_run
    /// use gql_parser::ParseOptions;
    ///
    /// let parsed = ParseOptions::new()
    ///     .unescape(true)
    ///     .parse(r"session set graph /`my\ngraph`");
    /// ```
    pub fn unescape(&mut self, unescape: bool) -> &mut Self {
        self.0.unescape = unescape;
        self
    }

    /// Parses a GQL query `gql` into a spanned abstract syntax tree with the options specified by
    /// `self`.
    ///
    /// # Errors
    /// This function will return an error if `gql` is not a valid GQL query. The error will carry
    /// fancy diagnostics if feature `miette` is enabled.
    ///
    /// Currently, we provide only simple and non-informative errors as defined in [`Error`]. More
    /// specific errors will be introduced in the future.
    ///
    /// # Examples
    /// ```
    /// use gql_parser::ParseOptions;
    ///
    /// let program = ParseOptions::new().parse("SESSION CLOSE");
    /// assert!(program.is_ok());
    /// assert_eq!(program.unwrap().span(), 0..13);
    /// ```
    pub fn parse(&self, gql: &str) -> Result<Spanned<Program>, Error> {
        let tokens = tokenize(gql)?;
        let stream = build_token_stream(&tokens, self.0.clone());
        gql_program
            .parse(stream)
            .map_err(|e| match tokens.get(e.offset()) {
                Some(token) => Error::unexpected(gql, token.span.clone()),
                None => Error::UnexpectedEof,
            })
    }
}

#[derive(Debug, Clone)]
pub(super) struct ParseOptionsInner {
    unescape: bool,
}

impl Default for ParseOptionsInner {
    fn default() -> Self {
        Self { unescape: true }
    }
}

impl ParseOptionsInner {
    pub(super) fn unescape(&self) -> bool {
        self.unescape
    }
}
