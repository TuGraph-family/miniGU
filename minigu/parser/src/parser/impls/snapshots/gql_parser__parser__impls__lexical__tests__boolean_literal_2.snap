---
source: minigu/parser/src/parser/impls/lexical.rs
expression: parsed
---
- "False"
- start: 0
  end: 5
